File: plugins/system-report/skills/system-report/templates/reporter.py
# =============================================================================
# system-report Reporter — TELEMETRY TUỲ CHỌN, MẶC ĐỊNH TẮT.
#
#   Gỡ bỏ hoàn toàn = xoá file này + 1 dòng wiring (`report.install()`). Hết. Không dấu vết.
#
# Bật:  REPORT_ENABLED=true  REPORT_INSTANCE={{INSTANCE}}  REPORT_REDIS_HOST=...
# Wiring (đúng 1 dòng, đặt ngay sau khi cấu hình logging xong):
#       import report; report.install()
#
# BẤT BIẾN (system-report §3.4):
#   · airtight  — nuốt MỌI exception, không bao giờ raise/blocking vào caller
#   · opt-in    — REPORT_ENABLED chưa bật thì install() là no-op tuyệt đối
#   · 1 file    — zero dependency, nói RESP thẳng qua TCP socket
#   · drop-safe — hàng đợi có trần; đầy thì VỨT log, không bao giờ chặn app
#   · chỉ ghi namespace REPORT_* trên 1 Redis-log RIÊNG (không phải Redis production)
# =============================================================================
import atexit
import logging
import os
import queue
import socket
import threading
import time
from datetime import datetime

_ENABLED = os.getenv("REPORT_ENABLED", "false").lower() in ("1", "true", "yes")
_INSTANCE = os.getenv("REPORT_INSTANCE") or socket.gethostname()
_HOST = os.getenv("REPORT_REDIS_HOST", "127.0.0.1")
_PORT = int(os.getenv("REPORT_REDIS_PORT", "6399") or 6399)
_DB = os.getenv("REPORT_REDIS_DB", "0")
_PASS = os.getenv("REPORT_REDIS_PASS", "")
_TTL = int(os.getenv("REPORT_TTL_DAYS", "7") or 7) * 86400
_FLUSH_SEC = float(os.getenv("REPORT_FLUSH_SEC", "5") or 5)
_QUEUE_MAX = int(os.getenv("REPORT_QUEUE_MAX", "2000") or 2000)
_LEVEL = os.getenv("REPORT_LEVEL", "WARNING").upper()

_q: "queue.Queue[str]" = queue.Queue(maxsize=_QUEUE_MAX)
_dropped = 0
_started = False
_lock = threading.Lock()


# ── RESP tối giản (không cần thư viện redis) ─────────────────────────────────
class _Conn:
    def __init__(self):
        self.sock = None
        self.rf = None

    def connect(self):
        self.close()
        s = socket.create_connection((_HOST, _PORT), timeout=5)
        s.settimeout(5)
        self.sock, self.rf = s, s.makefile("rb")
        if _PASS:
            self.cmd("AUTH", _PASS)
        if _DB and _DB != "0":
            self.cmd("SELECT", _DB)

    def cmd(self, *args):
        buf = bytearray(b"*%d\r\n" % len(args))
        for a in args:
            b = a.encode("utf-8", "replace") if isinstance(a, str) else a
            buf += b"$%d\r\n" % len(b) + b + b"\r\n"
        self.sock.sendall(bytes(buf))
        return self._reply()

    def _reply(self):
        line = self.rf.readline()
        if not line:
            raise IOError("redis closed")
        tag, body = line[:1], line[1:-2]
        if tag in b"+:-":
            return body
        if tag == b"$":
            n = int(body)
            if n < 0:
                return None
            data = self.rf.read(n + 2)
            return data[:-2]
        if tag == b"*":
            return [self._reply() for _ in range(max(0, int(body)))]
        return body

    def close(self):
        try:
            if self.rf:
                self.rf.close()
            if self.sock:
                self.sock.close()
        except Exception:
            pass
        self.sock = self.rf = None


def _key():
    return "REPORT_DAILY:%s:%s" % (datetime.now().strftime("%Y-%m-%d"), _INSTANCE)


def _worker():
    conn, registered = _Conn(), False
    while True:
        try:
            first = _q.get()  # chặn worker (thread nền), KHÔNG chặn app
        except Exception:
            return
        batch = [first]
        deadline = time.monotonic() + _FLUSH_SEC
        while len(batch) < 200:
            left = deadline - time.monotonic()
            if left <= 0:
                break
            try:
                batch.append(_q.get(timeout=left))
            except Exception:
                break
        try:
            if conn.sock is None:
                conn.connect()
                registered = False
            if not registered:
                conn.cmd("SADD", "REPORT_INSTANCES", _INSTANCE)
                registered = True
            payload = "\n".join(batch) + "\n"
            k = _key()
            conn.cmd("APPEND", k, payload)
            conn.cmd("EXPIRE", k, str(_TTL))
        except Exception:
            conn.close()  # lần sau tự kết nối lại; log của batch này bị bỏ — chấp nhận
        finally:
            for _ in batch:
                try:
                    _q.task_done()
                except Exception:
                    pass


def log(line: str) -> None:
    """Đẩy 1 dòng vào báo cáo ngày. Không bao giờ raise, không bao giờ block."""
    global _dropped
    if not _ENABLED:
        return
    try:
        ts = datetime.now().strftime("%H:%M:%S")
        _q.put_nowait("%s [%s] %s" % (ts, _INSTANCE, str(line).replace("\r", " ")))
    except queue.Full:
        _dropped += 1
    except Exception:
        pass


class _Handler(logging.Handler):
    def emit(self, record):
        try:
            log(self.format(record))
        except Exception:
            pass


def install() -> None:
    """1 dòng wiring duy nhất. Chưa bật REPORT_ENABLED → no-op tuyệt đối."""
    global _started
    if not _ENABLED:
        return
    try:
        with _lock:
            if _started:
                return
            _started = True
        t = threading.Thread(target=_worker, name="system-report", daemon=True)
        t.start()
        h = _Handler(level=getattr(logging, _LEVEL, logging.WARNING))
        h.setFormatter(logging.Formatter("%(levelname)s %(name)s: %(message)s"))
        logging.getLogger().addHandler(h)
        atexit.register(lambda: log("instance stopping (dropped=%d)" % _dropped))
        log("instance started")
    except Exception:
        pass  # telemetry hỏng KHÔNG được ảnh hưởng hệ thống

File: plugins/system-report/skills/system-report/templates/reporter.ts
// =============================================================================
// system-report Reporter — TELEMETRY TUỲ CHỌN, MẶC ĐỊNH TẮT.
//
//   Gỡ bỏ hoàn toàn = xoá file này + 1 dòng wiring (`installReporter()`). Hết.
//
// Bật:  REPORT_ENABLED=true  REPORT_INSTANCE={{INSTANCE}}  REPORT_REDIS_HOST=...
// Wiring (đúng 1 dòng, đặt sớm nhất trong entrypoint):
//       import { installReporter } from "./reporter"; installReporter();
//
// BẤT BIẾN (system-report §3.4):
//   · airtight  — nuốt MỌI exception; không throw, không await, không chặn event loop
//   · opt-in    — REPORT_ENABLED chưa bật thì installReporter() là no-op tuyệt đối
//   · 1 file    — zero dependency, nói RESP thẳng qua node:net
//   · drop-safe — hàng đợi có trần; đầy thì VỨT log
//   · chỉ ghi namespace REPORT_* trên 1 Redis-log RIÊNG (không phải Redis production)
// =============================================================================
import net from "node:net";
import os from "node:os";

const env = process.env;
const ENABLED = /^(1|true|yes)$/i.test(env.REPORT_ENABLED ?? "false");
const INSTANCE = env.REPORT_INSTANCE || os.hostname();
const HOST = env.REPORT_REDIS_HOST || "127.0.0.1";
const PORT = Number(env.REPORT_REDIS_PORT || 6399);
const DB = env.REPORT_REDIS_DB || "0";
const PASS = env.REPORT_REDIS_PASS || "";
const TTL = Number(env.REPORT_TTL_DAYS || 7) * 86400;
const FLUSH_MS = Number(env.REPORT_FLUSH_SEC || 5) * 1000;
const QUEUE_MAX = Number(env.REPORT_QUEUE_MAX || 2000);

let queue: string[] = [];
let dropped = 0;
let started = false;
let sock: net.Socket | null = null;
let connecting = false;
let registered = false;

function resp(...args: string[]): Buffer {
  const parts = [Buffer.from(`*${args.length}\r\n`)];
  for (const a of args) {
    const b = Buffer.from(String(a), "utf8");
    parts.push(Buffer.from(`$${b.length}\r\n`), b, Buffer.from("\r\n"));
  }
  return Buffer.concat(parts);
}

function drop(): void {
  try {
    sock?.destroy();
  } catch {
    /* ignore */
  }
  sock = null;
  registered = false;
}

function connect(): void {
  if (sock || connecting) return;
  connecting = true;
  try {
    const s = net.createConnection({ host: HOST, port: PORT });
    s.setNoDelay(true);
    s.unref(); // không giữ process sống chỉ vì telemetry
    s.on("connect", () => {
      connecting = false;
      sock = s;
      try {
        if (PASS) s.write(resp("AUTH", PASS));
        if (DB !== "0") s.write(resp("SELECT", DB));
      } catch {
        drop();
      }
    });
    s.on("data", () => {
      /* nuốt reply, tránh phình buffer */
    });
    s.on("error", () => {
      connecting = false;
      drop();
    });
    s.on("close", () => {
      connecting = false;
      drop();
    });
  } catch {
    connecting = false;
    drop();
  }
}

function today(): string {
  const d = new Date();
  const p = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())}`;
}

function flush(): void {
  if (queue.length === 0) return;
  connect();
  if (!sock || !sock.writable) return; // giữ hàng đợi, thử lại lượt sau
  const batch = queue;
  queue = [];
  try {
    const key = `REPORT_DAILY:${today()}:${INSTANCE}`;
    if (!registered) {
      sock.write(resp("SADD", "REPORT_INSTANCES", INSTANCE));
      registered = true;
    }
    sock.write(resp("APPEND", key, batch.join("\n") + "\n"));
    sock.write(resp("EXPIRE", key, String(TTL)));
  } catch {
    drop(); // batch này mất — chấp nhận, telemetry không được làm phiền app
  }
}

/** Đẩy 1 dòng vào báo cáo ngày. Không bao giờ throw, không bao giờ block. */
export function report(line: string): void {
  if (!ENABLED) return;
  try {
    if (queue.length >= QUEUE_MAX) {
      dropped++;
      return;
    }
    const d = new Date();
    const p = (n: number) => String(n).padStart(2, "0");
    const t = `${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`; // giờ LOCAL, khớp key ngày
    queue.push(`${t} [${INSTANCE}] ${String(line).replace(/\r/g, " ")}`);
  } catch {
    /* ignore */
  }
}

/** 1 dòng wiring duy nhất. Chưa bật REPORT_ENABLED → no-op tuyệt đối. */
export function installReporter(): void {
  if (!ENABLED || started) return;
  started = true;
  try {
    const timer = setInterval(flush, FLUSH_MS);
    timer.unref();

    for (const level of ["error", "warn"] as const) {
      const orig = console[level].bind(console);
      console[level] = (...a: unknown[]) => {
        report(`${level.toUpperCase()} ${a.map(String).join(" ")}`);
        orig(...(a as []));
      };
    }
    process.on("uncaughtException", (e) => report(`FATAL uncaughtException: ${e?.stack || e}`));
    process.on("unhandledRejection", (e) => report(`FATAL unhandledRejection: ${String(e)}`));
    process.on("exit", () => {
      report(`instance stopping (dropped=${dropped})`);
      flush();
    });
    report("instance started");
    flush();
  } catch {
    /* telemetry hỏng KHÔNG được ảnh hưởng hệ thống */
  }
}
